nav2_waypoint_follower/include/nav2_waypoint_follower/waypoint_follower.hpp

```diff
@@ -100,22 +112,78 @@ class WaypointFollower : public nav2_util::LifecycleNode
    */
   nav2_util::CallbackReturn on_shutdown(const rclcpp_lifecycle::State & state) override;
 
+  /**
+   * @brief Templated function to perform internal logic behind waypoint following,
+   *        Both GPS and non GPS waypoint following callbacks makes use of this function when a client asked to do so.
+   *        Callbacks fills in appropriate types for the tempelated types, see followWaypointCallback funtions for details.
+   *
+   * @tparam T action_server
+   * @tparam V feedback
+   * @tparam Z result
+   * @param action_server
+   * @param poses
+   * @param feedback
+   * @param result
+   */
+  template<typename T, typename V, typename Z>
+  void followWaypointsHandler(const T & action_server, const V & feedback, const Z & result);
+
   /**
    * @brief Action server callbacks
    */
-  void followWaypoints();
+  void followWaypointsCallback();
 
   /**
-   * @brief Action client result callback
-   * @param result Result of action server updated asynchronously
+   * @brief send robot through each of GPS
+   *        point , which are converted to map frame first then using a client to
+   *        `FollowWaypoints` action.
+   *
+   * @param waypoints, acquired from action client
    */
-  void resultCallback(const rclcpp_action::ClientGoalHandle<ClientT>::WrappedResult & result);
+  void followGPSWaypointsCallback();
+
+  /**
+ * @brief Action client result callback
+ * @param result Result of action server updated asynchronously
+ */
+  template<typename T>
+  void resultCallback(const T & result);
 
   /**
    * @brief Action client goal response callback
    * @param goal Response of action server updated asynchronously
    */
-  void goalResponseCallback(const rclcpp_action::ClientGoalHandle<ClientT>::SharedPtr & goal);
+  template<typename T>
+  void goalResponseCallback(const T & goal);
+
+  /**
+   * @brief given some gps_poses, converts them to map frame using robot_localization's service `fromLL`.
+   *        Constructs a vector of stamped poses in map frame and returns them.
+   *
+   * @param gps_poses, from the action server
+   * @return std::vector<geometry_msgs::msg::PoseStamped>
+   */
+  std::vector<geometry_msgs::msg::PoseStamped> convertGPSPosesToMapPoses(
+    const std::vector<geographic_msgs::msg::GeoPose> & gps_poses);
+
+
+  /**
+   * @brief get the latest poses on the action server goal. If they are GPS poses,
+   * convert them to the global cartesian frame using /fromLL robot localization
+   * server
+   *
+   * @param action_server, to which the goal was sent
+   * @return std::vector<geometry_msgs::msg::PoseStamped>
+   */
+  template<typename T>
+  std::vector<geometry_msgs::msg::PoseStamped> getLatestGoalPoses(const T & action_server);
+
+  // Common vars used for both GPS and cartesian point following
+  rclcpp::Node::SharedPtr client_node_;
```

Comment:
Remove this, there should be no client nodes.

Reply:
Done